torch/csrc/distributed/rpc/rref_impl.cpp

```diff
@@ -143,6 +152,38 @@ IValue UserRRef::toHere() {
   }
 }
 
+RRefForkData UserRRef::fork() const {
+  // Note [Best-Effort Check on Deleted UserRRefs]
+  // ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+  // This check does not guarantee correctness, as there could be another thread
+  // trying to delete this UserRRef concurrently. Passing this check does not
+  // mean this RRef will be alive throughout this function. This is just our
+  // best-effort attempt to raise proper error messages. The behavior of using
+  // deleted UserRRefs is undefined.
+  //
+  // The reason for not implementing strict checks are:
+  // 1. This would need to acquire lock on deletedOnOwnerMutex_, which would
+  //    introduce unnecessary overhead for most normal use cases.
+  // 2. This would introduce a lot of complexities to get the behavior correct.
+  //    Assume we acquired the lock here, and there is another thread X block
+  //    waiting in tryDel() on the lock. Exiting this fork function would
+  //    unblock thread X. However, while X proceeds with deleting this UserRRef,
+  //    the call site of fork() might have added the UserRRef to
+  //    pendingChildren_ map, but up to this point, nothing prevents X from
+  //    deleting this RRef even if it shouldn't do so due to the state change
+  //    in pendingChildren_. We might be able to get it right for now by locking
+  //    and checking pendingChildren_ in X, but the complexity does not seem to
+  //    worth the gain.
+  TORCH_CHECK(
+      !deletedOnOwner_,
+      "User RRef with RRefId=",
+      rrefId(),
+      " and ForkId=",
+      forkId(),
+      " has been deleted. Cannot call fork an UserRRef after deletion.");
+  return RRef::fork();
+}
```

Comment:
Not asking for changes, but what could be a robust way to avoid forking on graceful shutdown. A similar thing to agent.sync() to flush all tasks in the queues on all workers?

Reply:
Yep, it would at least require a global flushing, which we are trying to get rid of. Another thing is we don't have anything to prevent applications to create a local thread and use RRef/RPC in that thread. So it possible that the main thread is calling shutdown, but another thread is fork an RRef concurrently?